Update the documentation guide within the development guide

namurphy opened the pull request:
The goal of this PR is to expand the documentation within our development guide on how to write documentation in order to make it more complete and accessible to new contributors.  This is in preparation for [Plasma Hack Week](https://hack.plasmapy.org).

Here's an incomplete list of things I'm planning.  I'll probably edit this list as we go.

 - [x] Introduce ReST, including examples
 - [x] Briefly introduce Markdown
 - [x] Introduce Sphinx
 - [x] Discuss the numpydoc standard
 - [x] Add an example docstring with links, references, math, etc.  
 - [x] Discuss ReST substitutions (should merge #1147 first, and then incorporate it into here)
 - [x] Mention `plasmapy_sphinx` and create an issue on adding more information on it to the documentation guide
 - [x] Update the description of how to build documentation

rocco8773 commented on docs/development/doc_guide.rst:
```diff
@@ -60,17 +343,4 @@ Narrative Documentation
-    we can write |Particle|.
-
-Substitutions should be defined when they are used in multiple files.
+.. replace:: _`Read the Docs`: https://readthedocs.org/
```
```suggestion
.. _`Read the Docs`: https://readthedocs.org/
```

This is what is causing the current failed check.

namurphy replied:
Good catch — thanks!

rocco8773 commented on docs/development/doc_guide.rst:
```diff
@@ -1,17 +1,164 @@
+requirements and guidelines to be followed during the development of
+PlasmaPy and affiliated packages.
+
+PlasmaPy's online documentation is hosted by `Read the Docs
+<https://readthedocs.org/>`_ and can be found online at:
```
```suggestion
PlasmaPy's online documentation is hosted by `Read the Docs`_
and can be found online at:
```

rocco8773 commented on docs/development/doc_guide.rst:
```diff
@@ -1,17 +1,164 @@
+PlasmaPy's online documentation is hosted by `Read the Docs`_
+and can be found online at:
+
+* Most recent stable release:
+  `https://docs.plasmapy.org <https://docs.plasmapy.org>`_ or
+  `https://docs.plasmapy.org/en/stable/ <https://docs.plasmapy.org/en/stable/>`_
+
+* Latest version on GitHub:
+  `https://docs.plasmapy.org/en/latest/ <https://docs.plasmapy.org/en/latest/>`_
```
Might be useful to define the difference between what is meant by stable and latest, i.e. stable = last release and latest = Head of master.

rocco8773 replied:
This might also be worth adding in an admonition on the documentation landing page.  I would not be surprised if this is a common point of confusion when users seek out documentation, any documentation not just PlasmaPy.

namurphy replied:
Yeah, an admonition would be really helpful.  I created #1164 for doing this on the landing page.

namurphy commented:
- [ ] Briefly mention & link to page on [Sphinx templating](https://www.sphinx-doc.org/en/master/templating.html) with jinja2

namurphy commented on docs/development/doc_guide.rst:
```diff
@@ -1,76 +1,606 @@
+
+.. code-block:: bash
+
+   pip install tox -r requirements.txt
```
Would it make sense to add `tox` to `requirements/extras.txt`?  We do end up using it for development quite a lot, and it's also not a large package.

rocco8773 replied:
Why are you using `tox` here?  This can be done by just doing `pip install -r requirements.txt`. `tox` is not needed.

rocco8773 replied:
With that said, yes `tox` should be added to the `extras.txt`.  Now I see why `tox` is in this code example.

I say add, `tox` to `extras.txt` and remove it from the code-block here.

rocco8773 commented on docs/development/doc_guide.rst:
```diff
@@ -1,76 +1,633 @@
+  found at `https://docs.plasmapy.org/en/latest/
+  <https://docs.plasmapy.org/en/latest/>`_.
+
+A preview of the documentation is generated every time a pull request
+is created or updated. You can access this preview by scrolling down
+to the checks at the bottom of a pull request, and clicking on
+``Details`` next to ``docs/readthedocs.org:plasmapy``.
```
I like this bit, but it could be made clearer by adding a screenshot from an actual PR.

namurphy replied:
That's a good idea for a first contribution.  I'll create an issue for it.

rocco8773 commented on docs/development/doc_guide.rst:
```diff
@@ -1,76 +1,633 @@
+
+   pip install tox -r requirements.txt
+
+You can use `tox`_ to build the documentation from within the main
```
A little foreshadowing here would be a could thing.  You should specifically say there are two method to build the documentation: using `tox` and using `make`.  Then maybe give a brief explanation why you would use one over the other.

namurphy replied:
Could you say a bit more about the reasons to use `tox` vs. `make`?  I've mostly used `tox` since `make` brings back bad memories of trying to compile Fortran code.

rocco8773 replied:
The big difference, `make` will just look at the current directory structure and build the documentation off of that.  This means you have to have properly setup all your dependencies ahead of time in whatever environment you are working in.  `tox` is a lot more computationally intensive since it will build the package and environment before building the documentation.  `tox` is great for CIs but I absolutely hate it for local builds because it does so much un-needed extra bits.

rocco8773 commented on docs/development/doc_guide.rst:
```diff
@@ -1,76 +1,633 @@
+
+.. code-block:: bash
+
+   make html
```
Make a mention that there is a `make clean` (probably something similar for `tox`) that allows you to remove all build files.  This is often need when you make a change to a file that does not trigger sphinx to rebuild that file you altered.  This happen more frequently than one would expect.

rocco8773 commented on docs/development/doc_guide.rst:
```diff
@@ -1,76 +1,633 @@
+markup language. reST is human readable when viewed within a
+source code file or when printed out using `help`. reST also contains
+markup that allows the text to be transformed into `PlasmaPy's online
+documentation`_. reST files end in ``.rst``. Documentation contained
+within ``.py`` files is written in reST.
```
I'm not sure how to word this, but make some kind of indication that docstrings are the documentation source within `.py` files.

rocco8773 commented on docs/development/doc_guide.rst:
```diff
@@ -1,76 +1,633 @@
 
-* use `Tox <https://tox.readthedocs.io/en/latest/>`_ with ``tox -e build_docs`` from within the main PlasmaPy repository directory, or
-* enter the ``docs`` directory and run ``make html``.
+Here we show some examples of reST that are commonly used in PlasmaPy.
```
Add a comment to the effect of "..., but please refer to the reST and sphinx documentation for a more complete list of available directives and syntax."

rocco8773 commented on docs/development/doc_guide.rst:
```diff
@@ -1,76 +1,633 @@
+
+We can link to code objects by enclosing them in back ticks.
+
+.. code-block:: rst
+
+  Here is a reference to `plasmapy.particles` that will write out the
+  full namespace when `sphinx` generates the documentation and generates
+  the link. Only the word "Particle" will show up if we prepend a
+  tilde like in `~plasmapy.particles.particle_class.Particle`.
```
Maybe add the render of this example after the code block.

rocco8773 commented on docs/development/doc_guide.rst:
```diff
@@ -1,76 +1,633 @@
+  the link. Only the word "Particle" will show up if we prepend a
+  tilde like in `~plasmapy.particles.particle_class.Particle`.
+
+This linking will work for `python` commands as well as commonly used
+packages like `numpy`, `astropy`, `scipy`, and `pandas`. Intersphinx_
+mappings are defined in the ``intersphinx_mapping`` variable in
+`docs/conf.py`_.
```
```suggestion
This linking will work for `python` commands as well as commonly used
packages like `numpy`, `astropy`, `scipy`, and `pandas`.   This cross-package
linking is made possible with the `sphinx.ext.intersphinx` extension and
the sources used for this cross-linking is defined in the
``intersphinx_mapping`` variable in `docs/conf.py`_.  If a cross-link is not
working as expected this is usually due to (1) a typo, (2) the package is
not defined in the ``intersphinx_mapping`` configuration value, or (3)
the referenced source package does not properly or fully index their
own code (this is common in open-source packages).
```

rocco8773 replied:
@namurphy Why did you mark this as resolved?  What is your take on this?

namurphy replied:
Huh...I have no memory of marking this or other comments as resolved.  It could have been that I made changes to those lines of code which marked these comments as "outdated"...or possibly that I forgot how to use GitHub.  Whatever the case, I liked this suggestion.

rocco8773 commented on docs/development/doc_guide.rst:
```diff
@@ -1,76 +1,633 @@
+
+`sphinx` can format code blocks for Python and the Python console.
+
+   .. code-block:: rst
+
+      .. code-block:: python
+
+         def sample_function():
+             return 42
+
+      .. code-block:: pycon
+
+         >>> print(6 * 9)
+         54
```
Again, include the rendered version of the example.  I really wish I knew of a clean why placing the code and rendered examples side-by-side.  That would make it more readable.

rocco8773 commented on docs/development/doc_guide.rst:
````diff
@@ -1,76 +1,633 @@
+
+      \alpha = \beta + \gamma
+
+Math can be in-line, like ``:math:\`x\```. Using Unicode characters
+makes math like ``:math:\`α + β + γ\``` easier to read in source code.
````
```suggestion
Math can be in-line, like `` :math:`x` ``. Using Unicode characters
makes math like `` :math:`α + β + γ` `` easier to read in source code.
```

The backslashes are showing up on RTD.  I think my suggestion will fix this.

rocco8773 replied:
Again, I don't know why this was resolved.

namurphy replied:
Yep, this suggestion needs to be implemented too.

rocco8773 commented on docs/development/doc_guide.rst:
```diff
@@ -1,76 +1,633 @@
+Sphinx
+------
+
+`sphinx` is the software used to generate `PlasmaPy's online documentation`_
+from reST files and Python docstrings.
```
```suggestion
`sphinx` is the software used to generate `PlasmaPy's online documentation`_
from reST files and Python docstrings.  It was originally created to write Python's
documentation and has become the de facto software for documenting python
packages.  Almost all python open-source packages utilize `sphinx` to generate
their documentation.
```

rocco8773 commented on docs/conf.py:
```diff
@@ -54,7 +54,8 @@
 # Add any Sphinx extension module names here, as strings. They can be
 # extensions coming with Sphinx (named 'sphinx.ext.*') or your custom
-# ones.
+# ones. When extensions are removed or added, please update the
+# documentation guide in the development guide.
```
Reference the exact file and section that should be updated, so a user does not have to figure that out on their own.

rocco8773 commented on docs/conf.py:
```diff
@@ -70,6 +71,9 @@
     "plasmapy_sphinx",
 ]
 
+# Intersphinx generates automatic links to the documentation of objects
+# in other packages. When mappings are removed or added, please update
+# the documentation guide in the development guide.
```
Reference the exact file and section that should be updated, so a user does not have to figure that out on their own.

rocco8773 commented on docs/development/doc_guide.rst:
```diff
@@ -1,76 +1,633 @@
+  Note that `plasmapy_sphinx` is expected to be broken out into its own
+  package at a future date.
+
+These extensions are specified in ``extensions`` in `docs/conf.py`_.
```
```suggestion
These extensions are specified in :confval:`extensions` configuration
value in `docs/conf.py`_.
```

This should cross-link to the ``extensions`` documentation in sphinx.

rocco8773 replied:
Actually, you should be able to do this with the `intersphinx_mappings` as well, ``:confvl:`intersphinx_mappings` ``.

rocco8773 commented on docs/development/doc_guide.rst:
```diff
@@ -1,76 +1,633 @@
+References to other packages
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Intersphinx_ allows the automatic generation of links to the
+documentation of objects in other projects. The mappings are defined in
+the ``intersphinx_mapping`` dictionary in `docs/conf.py`_, and include
+`python`, `numpy`, `scipy`, `astropy`, `pandas`, `sphinx`, and
+`sphinx_automodapi`.
```
```suggestion
Intersphinx_ allows the automatic generation of links to the
documentation of objects in other projects. The mappings are defined in
the :confval:`intersphinx_mapping` dictionary in `docs/conf.py`_, and include
`python`, `numpy`, `scipy`, `astropy`, `pandas`, `sphinx`, and
`sphinx_automodapi`.
```

What I said in https://github.com/PlasmaPy/PlasmaPy/pull/1156/files?file-filters%5B%5D=.rst#r654641786.  Should propagate this to all references of `intersphinx_mappings`.

rocco8773 commented on docs/development/doc_guide.rst:
```diff
@@ -1,76 +1,633 @@
+`python`, `numpy`, `scipy`, `astropy`, `pandas`, `sphinx`, and
+`sphinx_automodapi`.
+
+When we include ``astropy.units.Quantity`` in the documentation, it
+will show up as `astropy.units.Quantity` with a link to Astropy's
+documentation.
```
Give the `~` example as well.

rocco8773 commented on docs/development/doc_guide.rst:
```diff
@@ -1,76 +1,633 @@
+For an up-to-date list of substitutions, please refer to the
+`docs/common_links.rst`_ file.
+
+Because substitutions are performed when `sphinx` builds the
+documentation, they will not be performed before `help` accesses the
+docstring of an `object`. For example, when ``|Particle|`` is used in
```
```suggestion
Since substitutions are performed by `sphinx` when the documentation
is built, any substitution used in docstrings will not show up when using
Python's `help` function (or the like). For example, when ``|Particle|`` is used in
```

rocco8773 commented on docs/development/doc_guide.rst:
```diff
@@ -1,76 +1,633 @@
-==========
+----------
+
+A docstring is a comment at the beginning of a function or another
+object that provides information on how to use that function.
+Docstrings begin with ``r"""`` (required when including backslashes,
+such as using LaTeX code in equations) or ``"""``, and end with
+``"""``.
```
This would be a good spot to reference Python's PEP on docstrings, https://www.python.org/dev/peps/pep-0257/

rocco8773 commented on docs/development/doc_guide.rst:
```diff
@@ -1,76 +1,633 @@
+
+  def sample_function():
+      r"""
+      Return ...
```
```suggestion
      Readable description of sample_function.
```

I don't want to encourage people to start of a docstring with "Return."  It redundant, just say what the functionality is doing.

namurphy replied:
Huh.  I got that wording from one of the recommendations in [PEP 257](https://www.python.org/dev/peps/pep-0257/#id16).

rocco8773 replied:
Yeah, I know.  It's widely used and in the PEP so I've never mentioned it.  You started removing them in a recent documentation refactor, so I'm hopeful that it gets propagated throughout the documentation.

Technically there's nothing wrong with it, but it's so redundant.  It doesn't do anything to expand the meaning of the documentation.  I already know a function is returning something, why are you telling me again!?  I have to read two extra words before I even get to the meat of the description.  It's just a pet-peeve of mine.  I won't use it but I also won't go throughout the documentation to strike it down.

rocco8773 commented on docs/development/doc_guide.rst:
```diff
@@ -1,76 +1,633 @@
-* Simple private functions may need only a one-line docstring.
+   def subtract(a, b, *, switch_order=False):
+       r"""
+       Return the difference between two integers.
```
```suggestion
       Compute the difference between two integers.
```

Just say what the function is doing, we already know it's returning something. This is probably the thing that annoys me the most about our docstrings and I'd like to discourage its usage.

rocco8773 commented on docs/development/doc_guide.rst:
```diff
@@ -1,76 +1,633 @@
+Documentation guidelines
+========================
+
+This section contains guidelines and best practices for writing
+PlasmaPy documentation.
```
Add a bullet stating that if the main docstring should only be 2-4 sentences followed by the "Parameters" section.  If more detail needs to be added, then the "Notes" section should be used to expand on the main component of the docstring.

rocco8773 commented on docs/development/doc_guide.rst:
```diff
@@ -1,76 +1,633 @@
+  * Represent names and acronyms for a software package as they are
+    represented in the documentation for that package.
+
+* Write the full namespace when referring to code objects within
+  PlasmaPy. For example, write
```
```suggestion
* When referencing PlasmaPy functionality, write the full namespace path to
  where the functionality is defined, not where it's conveniently accessed.
  For example, write
```

rocco8773 commented on docs/development/doc_guide.rst:
```diff
@@ -1,76 +1,633 @@
+* Write the full namespace when referring to code objects within
+  PlasmaPy. For example, write
+  ``~plasmapy.formulary.parameters.Alfven_speed`` rather than
+  ``~plasmapy.formulary.Alfven_speed``.  This does not need to be done
+  for intersphinx links to other packages; it is fine to do
+  ``~astropy.units.Quantity`` and not necessary to do
+  ``~astropy.units.quantity.Quantity``.
```
```suggestion
  ``~plasmapy.formulary.Alfven_speed``.  This does not necessarily
  need to be done when referencing external packages, since
  each package may have their own standard.  For example, Astropy's
  `~astropy.units.Quantity` class is defined in
  ``astropy.units.quantity.Quantity`` but is also indexed at
  ``~astropy.units.Quantity`` so either option will link to some
  documentation.
```

rocco8773 commented on docs/development/doc_guide.rst:
```diff
@@ -1,76 +1,633 @@
+  ``~astropy.units.Quantity`` and not necessary to do
+  ``~astropy.units.quantity.Quantity``.
+
+* For readability, limit documentation line lengths to ≲ 72 characters.
+  Longer line lengths may be used when necessary (e.g., for hyperlinks).
```
```suggestion
* For readability, limit documentation line lengths to ≲ 80 characters.
  Longer line lengths may be used when necessary (e.g., for hyperlinks).
```

This number comes from Python's PEP8 rule based on a max line length of 80.  Our max line length is 90 so I say we bump this up by 8-10 characters as well.

namurphy replied:
I looked up recently what the [optimal line length](https://baymard.com/blog/line-length-readability) is for readability and I've seen ∼50–60 characters as well as ∼50–75 characters.  Perhaps we could say ≲ 72 characters excluding indentations.  That'd typically be a max line length of 76 characters for regular functions, so it ends up being an average of our two preferences.

rocco8773 commented on docs/development/doc_guide.rst:
```diff
@@ -1,76 +1,633 @@
+  public Application Programming Interface (API) must have a docstring
+  that follows the `numpydoc`_ standard.
+
+* The first line of the docstring for a function or method should begin
+  with a word like "Return", "Calculate", or "Compute" and end with a
+  period.
```
```suggestion
* The first line of the docstring for a function or method should begin
  with a word like "Calculate" or "Compute" and end with a period.
```

rocco8773 commented on docs/development/doc_guide.rst:
```diff
@@ -1,76 +1,633 @@
+  object may be a single line, and otherwise should be in `numpydoc`_
+  format.
+
+* Dunder methods (e.g., code objects like ``__init__`` that begin and
+  end with two underscores) only need to have docstrings when needed to
+  describe non-standard or potentially unexpected behavior.
```
I wouldn't use `__init__` or `__new__` as examples here, because they get added to the class level docstring when rendering documentation.  I think they should get their own bullet point.

State the dunder methods are not rendered in the online documentation and are not intended for the public API.  Docstrings for these methods should be written for the developer in mind.

rocco8773 commented on docs/development/doc_guide.rst:
```diff
@@ -1,76 +1,633 @@
+  end with two underscores) only need to have docstrings when needed to
+  describe non-standard or potentially unexpected behavior.
+
+* When an attribute in a class has both a getter (which is decorated
+  with `property`) and a setter method, both the getter and setter
+  should be documented in the docstring for the getter.
```
```suggestion
* When an attribute in a class has both a ``getter`` (which is decorated
  with `property`) and a ``setter`` decoration, then the ``getter`` and
  ``setter`` functionality should be documented in the docstring of
  method decorated with ``@property``.
```

rocco8773 commented on docs/development/doc_guide.rst:
```diff
@@ -1,76 +1,633 @@
+Narrative documentation guidelines
+----------------------------------
+
+* Each subpackage in PlasmaPy must have corresponding narrative
+  documentation.
```
```suggestion
* Each top-level subpackage in PlasmaPy must have corresponding narrative
  documentation.
```

I don't think this is even currently true for all sub-subpackages, and I'm okay with that.

rocco8773 commented on docs/development/doc_guide.rst:
```diff
@@ -1,76 +1,633 @@
+         def age(self, n):
+             self._age = n
+
+Narrative documentation guidelines
```
I would make some kind of comment that narrative documentation should be used when the full scope of some functionality can not be achieved with just docstrings of said functionality.  This often comes about when documentation needs to be more educational or there's a lot of connective tissue between complex functionality.

rocco8773 commented on docs/development/doc_guide.rst:
```diff
@@ -1,76 +1,633 @@
+* Use title case for page titles (e.g., "Title Case") and sentence case
+  for all other headings (e.g., "Sentence case").
+
+* When the narrative documentation does not reference a subpackage or
+  module file, it is necessary to create a stub file in
+  `docs/api_static`_ for that particular subpackage or module file.
+  Here are the sample contents for a stub file for
+  `plasmapy.particles.atomic``.  This file would be located at
+  ``docs/api_static/plasmapy.particles.atomic.rst``.
```
```suggestion
* When the narrative documentation does not index a subpackage
  (directory) or module (``.py`` file) with :dir:`automodule`,
  :dir:`automodapi`, or the like, then it is required to create a stub
  file for that particular subpackage or module in `doc/api_static`_ .
  Here are the sample contents for a stub file for
  `plasmapy.particles.atomic``.  This file would be located at
  ``docs/api_static/plasmapy.particles.atomic.rst``.
```

namurphy commented:
Thank you again for the review!  I don't have enough brain right now to be braining, but I'll try to implement your suggestions in the next few days.

namurphy commented:
This PR turned out to be a long one, but I suppose that this was necessary since this was a major rewrite.  

I believe I had addressed all of the comments from @rocco8773.  I'm still planning on checking the rendered documentation for this page more carefully, and will request a review after I finish making sure that things are showing up the way I'm intending them to, but that'll probably be another day.

rocco8773 commented on docs/whatsnew/0.1.0.rst:
```diff
@@ -34,8 +34,7 @@ New Features
 
-  - Automated documentation builds with `Sphinx
-    <http://www.sphinx-doc.org/>`_.
+  - Automated documentation builds with Sphinx.
```
How about adding

```rst
.. _Sphinx: https://www.sphinx-doc.org
```

to `docs/common_links.rst` so it can be used in places like here.

rocco8773 commented on docs/development/doc_guide.rst:
```diff
@@ -1,76 +1,829 @@
+`PlasmaPy's documentation`_ is hosted by `Read the Docs`_ and is
+available at these locations:
+
+* The documentation corresponding to the most recent official release is
+  labelled ``stable`` and is found at `https://docs.plasmapy.org/`_ or
+  `https://docs.plasmapy.org/en/stable/`_.
+
+* The documentation corresponding to the ``main`` branch on `PlasmaPy's
+  GitHub repository`_ is labelled ``latest`` and can be found at
+  `https://docs.plasmapy.org/en/latest/`_.
```
```suggestion
* The documentation corresponding to the most recent release to `PyPI`_ is
  labelled ``stable`` and is found at `https://docs.plasmapy.org/`_ or
  `https://docs.plasmapy.org/en/stable/`_.

* The documentation corresponding to the ongoing development on the
  ``main`` branch in `PlasmaPy's GitHub repository`_ , which is often ahead
  of the most recent release, is labelled ``latest`` and can be found at
  `https://docs.plasmapy.org/en/latest/`_.
```

I reworded this a little because I thought it would be useful to the reader to add a little more clarification about the difference between the most recent release and `main`.

rocco8773 commented on docs/development/doc_guide.rst:
```diff
@@ -1,76 +1,829 @@
-of Sphinx's website.
+There are two methods for building the documentation: make_ and tox_.
+
+* Using make_ will build the documentation based off of what is in the
+  current directory structure. Using make_ is quicker for local builds
+  but requires that setting up all dependencies ahead of time.
```
```suggestion
* Using make_ will build the documentation based off of what is in the
  current directory structure. It is quicker for local builds than tox_
  but requires you to install and setup all dependencies.
```

rocco8773 commented on docs/development/doc_guide.rst:
```diff
@@ -1,76 +1,829 @@
+* Using make_ will build the documentation based off of what is in the
+  current directory structure. Using make_ is quicker for local builds
+  but requires that setting up all dependencies ahead of time.
+* Using tox_ does not require setting up all dependencies ahead of time,
+  but is more computationally intensive since it builds the package and
+  environment before building the documentation. Consequently, PlasmaPy
+  uses tox_ for building the documentation on continuous integration
+  testing platforms.
```
```suggestion
* Using tox_ does not require setting up all dependencies ahead of time,
  but is more computationally intensive since it creates a virtual environment
  and builds the package before building the documentation. Consequently,
  PlasmaPy uses tox_ for building the documentation on continuous integration
  testing platforms.
```

rocco8773 commented on docs/development/doc_guide.rst:
```diff
@@ -1,76 +1,829 @@
+Including the ``-j n`` flag in the ``make`` command will enable a
+parallel build, where ``n`` is replaced with the number of processes.
+
+To remove all files from a documentation build, run:
```
```suggestion
To remove all files previously generated by ``make``, run:
```

Just so there's not confusion that `make clean` might remove EVERYTHING.

namurphy replied:
Good catch!

rocco8773 commented on docs/development/doc_guide.rst:
```diff
@@ -1,76 +1,829 @@
+
+   make clean
+
+This command is needed when you make a change to a file that does not
+trigger Sphinx_ to rebuild the file that you altered.
```
```suggestion
This command is needed when you make a change to a file that does not
trigger Sphinx_ to rebuild the file that you altered, for example modifying the CSS file.
```

rocco8773 commented on docs/development/doc_guide.rst:
```diff
@@ -1,76 +1,829 @@
-   The ``tox -e build_docs_no_examples`` command will build the documentation without
-   executing the :ref:`example notebooks <example_notebooks>`. It will also
-   pass with warnings.
+You can access the documentation landing page by opening
+:file:`docs/_build/html/index.html` with your browser of choice.
```
You should make a similar comment under the previous section, "Building documentation with make."

namurphy commented:
I'm really tempted to add an admonition that says...

```rst
.. danger::

   Sphinx rabbit holes often have dragons in them.
```

...but I'm afraid it won't make sense to people until it's far too late.

rocco8773 commented:
> 
> 
> I'm really tempted to add an admonition that says...
> 
> ```
> .. danger::
> 
>    Sphinx rabbit holes often have dragons in them.
> ```
> 
> ...but I'm afraid it won't make sense to people until it's far too late.

I like it and would approve the PR!!!

rocco8773 commented on docs/plasmapy_sphinx/__init__.py:
```diff
@@ -1,7 +1,6 @@
-This `Sphinx <https://www.sphinx-doc.org/en/master/>`_ extension package contains
-custom documentation functionality used to document
-`PlasmaPy's packages <https://github.com/PlasmaPy>`_ .
+This Sphinx_ extension package contains custom documentation functionality
+used to document `PlasmaPy's packages <https://github.com/PlasmaPy>`_ .
```
Can you revert this back?  I only say that so the Sphinx link is not broken when I spin this off into it's own package.  After I spin the package off I'll create a common links for that package.

rocco8773 commented on docs/development/doc_guide.rst:
```diff
@@ -1,76 +1,849 @@
+markup language. reST_ is human readable when viewed within a source
+code file or when printed out using `help`. reST_ also contains markup
+that allows the text to be transformed into `PlasmaPy's documentation`_.
+reST_ files end in :file:`.rst`. Documentation contained within
+:file:`.py` files are in the form of docstrings_, which are written in
+reST_.
```
```suggestion
reST_ files use the file extension :file:`.rst`. Documentation contained
within :file:`.py` files are in the form of docstrings_, which are written in
reST_.
```

namurphy replied:
I was wondering how to better word this.  Thanks!

rocco8773 commented on docs/development/doc_guide.rst:
```diff
@@ -1,76 +1,849 @@
+ReStructuredText Examples
+~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Here we show some examples of reST_ that are commonly used in PlasmaPy.
+Please refer to the documentation for Sphinx_ and reST_ for a list of
+available directives and syntax.
```
```suggestion
Here we show some examples of commonly used reST_ syntax in
PlasmaPy.  Please refer to the documentation for Sphinx_ and reST_ for
a list of available directives and syntax.
```

rocco8773 commented on docs/development/doc_guide.rst:
```diff
@@ -1,76 +1,849 @@
+the section on :ref:`external-references`. In-line code examples are
+enclosed in double back ticks.
+
+.. code-block:: rst
+
+   Here is a reference to `plasmapy.particles` that will show the full
+   namespace in the rendered documentation.
+
+   Only the word "Particle" will show up if a prepend a tilde like in
+   `~plasmapy.particles.particle_class.Particle`.
+
+   Here is some in-line code: ``import astropy.units as u``.
+
+This reST_ block will render as:
+
+   Here is a reference to `plasmapy.particles` that will show the full
+   namespace in the rendered documentation.
+
+   Only the word "Particle" will show up if a prepend a tilde like in
+   `~plasmapy.particles.particle_class.Particle`.
+
+   Here is some in-line code: ``import astropy.units as u``.
```
```suggestion
.. code-block:: rst

   Here `plasmapy.particles` provides a linked reference to the
   module's documentation.
   
   Adding a tilde at the beginning `~plasmapy.particles` still
   provides a linked reference to the associated documentation
   but shortens the display so only "particles" is displayed.

   Double backticks are used to show inline code that is not
   cross-referenced: ``import astropy.units as u``.

This reST_ block will render as:

   Here `plasmapy.particles` provides a linked reference to the
   module's documentation.
   
   Adding a tilde at the beginning `~plasmapy.particles` still
   provides a linked reference to the associated documentation
   but shortens the display so only "particles" is displayed.

   Double backticks are used to show inline code that is not
   cross-referenced: ``import astropy.units as u``.
```

I felt this could use a little more elaboration for clarity, so this is my take on it.  I mostly wanted to emphasize that a link to the documentation is generated.

rocco8773 commented on docs/development/doc_guide.rst:
```diff
@@ -1,76 +1,849 @@
+
+Here are some examples for linking to websites.
+
+.. code-block:: rst
+
+   Here is a link to `PlasmaPy's website <https://www.plasmapy.org>`_.
+
+   We can link to PlasmaPy's latest documentation_ or `Python's website`_.
+
+   .. _documentation: https://docs.plasmapy.org/en/latest/
+   .. _`Python's website`: https://www.python.org/
```
Is there a reason you did not include a rendered version here?

namurphy replied:
Yes, it was to avoid duplication with `common_links.rst`, at least for earlier commits.  With that said, yeah, it'd be worth changing the example so we can include a rendered version.

rocco8773 commented on docs/development/doc_guide.rst:
```diff
@@ -1,76 +1,849 @@
+.. code-block:: rst
+
+   An example of in-line math is :math:`x`. Using Unicode characters
+   like :math:`α + β + γ` makes math easier to read in the source code.
```
meh...it's a little subjective 🤣

rocco8773 commented on docs/development/doc_guide.rst:
```diff
@@ -1,76 +1,849 @@
+Markdown
+--------
+
+A few of PlasmaPy's files are written using Markdown_, such as README
+files and licenses from other packages. Markdown_ is simpler but more
+limited than reST_. Markdown_ files end with :file:`.md`. Posts on
+GitHub are written in `GitHub Flavored Markdown`_.  The following code
+block contains a few common examples of Markdown_ formatting.
```
```suggestion
A few of PlasmaPy's files are written using Markdown_, such as README
files and licenses from other packages. Markdown_ is simpler but more
limited than reST_. Markdown_ files use the file extension :file:`.md`. Posts on
GitHub are written in `GitHub Flavored Markdown`_.  The following code
block contains a few common examples of Markdown_ formatting.
```

rocco8773 commented on docs/development/doc_guide.rst:
```diff
@@ -1,76 +1,849 @@
+
+.. _external-references:
+
+References to other packages
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```
```suggestion
Cross-referencing external packages
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

Sphinx documentation uses the "cross-referencing" terminology in their documentation, so I feel it's best to be consistent with them.

rocco8773 commented on docs/development/doc_guide.rst:
```diff
@@ -1,76 +1,849 @@
+documentation of objects in other projects. This cross-package linking
+is made possible with the `sphinx.ext.intersphinx` extension.
+
+When we include ``astropy.units.Quantity`` in the documentation, it will
+show up as `astropy.units.Quantity` with a link to the appropriate page
+in Astropy documentation. Similarly, ``~astropy.units.Quantity`` will
+show up as `~astropy.units.Quantity`.
```
```suggestion
When we include `` `astropy.units.Quantity` `` in the documentation, it will
show up as `astropy.units.Quantity` with a link to the appropriate page
in Astropy documentation. Similarly, `` `~astropy.units.Quantity` `` will
show up as `~astropy.units.Quantity`.
```

I'm not 100% that this will work as intended so you'll have to test it out.  But, I think having the actual code syntax display here will make the documentation more clear to the reader.

rocco8773 commented on docs/development/doc_guide.rst:
```diff
@@ -1,76 +1,849 @@
+References to other packages
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Intersphinx_ allows the automatic generation of links to the
+documentation of objects in other projects. This cross-package linking
+is made possible with the `sphinx.ext.intersphinx` extension.
```
```suggestion
Intersphinx_ allows the automatic generation of links to the
documentation of objects in other projects. This cross-package linking
is made possible with the `sphinx.ext.intersphinx` extension and proper
package indexing by the external package using `sphinx.ext.autodoc`.
```

As you've discovered recently, proper cross-referencing is twofold: (1) us setting up `intersphinx` properly and (2) the desired package properly indexing their functionality.

rocco8773 commented on docs/development/doc_guide.rst:
```diff
@@ -1,76 +1,849 @@
+in Astropy documentation. Similarly, ``~astropy.units.Quantity`` will
+show up as `~astropy.units.Quantity`.
+
+The mappings are defined in the :confval:`intersphinx_mapping`
+dictionary in `docs/conf.py`_, and include Python_, NumPy_, SciPy_,
+Astropy_, pandas_, Sphinx_, and `sphinx_automodapi`_.
```
```suggestion
To make cross-referencing to an external package available its
mappings have to be defined in the :confval:`intersphinx_mapping`
configuration dictionary contained in `docs/conf.py`_.  PlasmaPy
has already include several packages like Python_, NumPy_, SciPy_,
Astropy_, Sphinx_, etc.
```

I wanted to make the last part of this statement a little more general so it does not get out of date immediately.  Plus I don't want to have to maintain it whenever we add a new package to the mappings.

namurphy replied:
I fully support future proofing documentation when possible.

rocco8773 commented on docs/development/doc_guide.rst:
```diff
@@ -1,76 +1,849 @@
+documentation. reST_ allows us to `define substitutions
+<https://docutils.sourceforge.io/docs/ref/rst/restructuredtext.html#substitution-definitions>`_.
+
+.. code-block:: rst
+
+   .. |Particle| replace:: `~plasmapy.particles.particle_class.Particle`
```
```suggestion
.. code-block:: rst

   .. |Particle| replace:: `~plasmapy.particles.particle_class.Particle`

Here whenever ``|Particle|`` is used Sphinx_ will replace it with 
`` `~plasmapy.particles.particle_class.Particle` `` during build time.
```

Nowhere in this section did you explicitly say Sphinx substitutes X for Y during building.  I wanted to directly say that.

rocco8773 commented on docs/development/doc_guide.rst:
```diff
@@ -1,76 +1,849 @@
+
+   .. |Particle| replace:: `~plasmapy.particles.particle_class.Particle`
+
+PlasmaPy has certain common substitutions pre-defined so that they can
+be used elsewhere in the documentation. For example, we can write
+``|Quantity|`` instead of ``~astropy.units.Quantity``, and
+``|Particle|`` instead of
+``~plasmapy.particles.particle_class.Particle``.  For an up-to-date list
+of substitutions, please refer to the `docs/common_links.rst`_ file.
+
+Since substitutions are performed by Sphinx_ when the documentation is
+built, any substitution used in docstrings will not show up when using
+Python's `help` function (or the like). For example, when ``|Particle|``
+is used in a docstring, `help` will show it as ``|Particle|`` rather
+than ``~plasmapy.particles.particle_class.Particle``.  Consequently,
+substitutions should not be used in docstrings when it is important that
+users have quick access to the full path of the `object` (such as in the
+``See Also`` section).
```
```suggestion
PlasmaPy has certain common substitutions pre-defined so that they can
be used elsewhere in the documentation. For example, we can write
``|Quantity|`` instead of ```~astropy.units.Quantity```, and
``|Particle|`` instead of
```~plasmapy.particles.particle_class.Particle```.  For an up-to-date list
of substitutions, please refer to the `docs/common_links.rst`_ file.

Since substitutions are performed by Sphinx_ when the documentation is
built, any substitution used in docstrings will not show up when using
Python's `help` function (or the like). For example, when ``|Particle|``
is used in a docstring, `help` will show it as ``|Particle|`` rather
than ```~plasmapy.particles.particle_class.Particle```.  Consequently,
substitutions should not be used in docstrings when it is important that
users have quick access to the full path of the `object` (such as in the
``See Also`` section).
```

rocco8773 commented on docs/development/doc_guide.rst:
```diff
@@ -1,76 +1,849 @@
+users have quick access to the full path of the `object` (such as in the
+``See Also`` section).
+
+Templating
+~~~~~~~~~~
+
+Sphinx_ uses the Jinja_ templating engine to generate HTML code. Jinja_
+may be used within the documentation when templating is necessary.
```
Should include a link to Sphinx's templating page, https://www.sphinx-doc.org/en/master/templating.html.

rocco8773 commented on docs/development/doc_guide.rst:
```diff
@@ -1,76 +1,849 @@
+----------
 
-* All public classes, methods, and functions should have docstrings.
+A docstring is a comment at the beginning of a function or another
+object that provides information on how to use that function.
+Docstrings begin with ``r"""`` (required when including backslashes,
+such as using LaTeX_ code in equations) or ``"""``, and end with
+``"""``.
```
```suggestion
A docstring_ is a comment at the beginning of a function or another
object that provides information on how to use that function
(see `PEP 257 <https://www.python.org/dev/peps/pep-0257/>`_).
Docstrings are designated by surrounding the content with
triple quotes ``"""This is my docstring."""``.  If a docstring
contains math that utilizes LaTex_ code, then we recommend
using a raw-string so you don't have to handle escape
characters on backslashes.  For example, the raw-string
``r""":math:`\alpha`"""`` will render the same as the normal string
``""":math:`\\alpha`"""``.
```

I've already noticed contributors using raw strings when it's not needed.  I think highlighting the difference between raw and normal strings here would be useful.

rocco8773 commented on docs/development/doc_guide.rst:
```diff
@@ -1,76 +1,849 @@
-* PlasmaPy uses the `numpydoc`_ standard for docstrings.
+In order to improve readability and maintain consistency, PlasmaPy uses
+the numpydoc_ standard for docstrings. Docstring conventions for Python_
+are more generally described in :pep:`257`.
```
I did not know there as a PEP role.  I like it!!

rocco8773 commented on docs/development/doc_guide.rst:
```diff
@@ -1,76 +1,849 @@
+  * Docstrings for private code objects do not get rendered in the
+    online documentation, and should be intended for contributors.
+
+* Dunder methods (e.g., code objects like ``__add__`` that begin and end
+  with two underscores) only need docstrings if it is necessary to
+  describe non-standard or potentially unexpected behavior. Custom
+  behavior associated with dunder methods should be described in the
+  class-level documentation.
+
+  * Docstrings for most dunder methods are not rendered in the online
+    documentation and should therefore be intended for contributors.
+
+  * Docstrings for ``__init__`` and ``__new__`` are included in the
+    class-level docstring when rendering documentation.
```
The `__call__` dunder is rendered for the documentation, so I'd add a bullet noting it's docstring should be written for the user and not just the contributor/developer.

rocco8773 commented on docs/development/doc_guide.rst:
```diff
@@ -1,76 +1,849 @@
+  * Docstrings for ``__init__`` and ``__new__`` are included in the
+    class-level docstring when rendering documentation.
+
+* When an attribute in a class has both a getter (which is decorated
+  with `property`) and a ``setter`` decoration, then the getter and
+  ``setter`` functionality should be documented in the docstring of the
+  attribute decorated with ``@property``.
```
```suggestion
* When an attribute in a class has both a getter (which is the method
  decorated with `@property`) and a ``setter`` decoration, then the
  getter and ``setter`` functionality should be documented in the
  docstring of the attribute decorated with ``@property``.
```

rocco8773 commented on docs/development/doc_guide.rst:
```diff
@@ -1,76 +1,849 @@
+* When the narrative documentation does not index a subpackage (a
+  directory) or module (a :file:`.py` file) with ``automodule``,
+  ``automodapi``, or the like, then it is required to create a stub file
+  for that particular subpackage or module in `docs/api_static`_ .  Here
+  are the sample contents for a stub file for
+  `plasmapy.particles.atomic`. This file would be located at jlkj;lj
+  :file:`docs/api_static/plasmapy.particles.atomic.rst`.
```
```suggestion
  for that particular subpackage or module in `docs/api_static`_ .  For
  example, the stub file for `plasmapy.particles.atomic` is placed at
  :file:`docs/api_static/plasampy.particles.atomic.rst` and its contents
  look like:
```

rocco8773 reviewed (changes requested):
This is looking good.  Just address the comments I made and I'll then be happy to approve the PR.
